### PR TITLE
Add AWS DynamoDB Scaler

### LINKED ISSUE #2482: AWS DynamoDb scaler
### Proposal

Scaler based on the result of a DynamoDB table query.

### Scaler Source

The single numeric result of a DynamoDb query.

### Scaling Mechanics

The scaling mechanics is based on the result of a table query in DynamoDB (similar to what has been done for MongoDB).

### Authentication Source

It can reuse the `identityOwner` and the `TriggerAuthentication`  implemented for the SQS Scaler.

### Anything else?

The idea is similar to the MongoDB scaler. It is possible to reuse the authentication bit already implemented for Amazon SQS.
DynamoDb paginates results over 1MB, see [https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Query.html#Query.Count](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Query.html#Query.Count)

### PR DESCRIPTION
**Pull request by samueleresca**
Signed-off-by: Samuele Resca <samuele.resca@gmail.com>

_Provide a description of what has been changed_

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [X] Tests have been added
- [X] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [X] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs/pull/680
- [X] Changelog has been updated

Fixes #2482

### CODE REVIEW

**Comment by tomkerkhove** [comment]
Thanks for your PR! Would you mind opening and linking a PR for the docs please?

File: pkg/scalers/aws_dynamodb_scaler.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+		Region:      aws.String(meta.awsRegion),
+		Credentials: creds,
+	})
+
+	return dbClient
+}
+
+func (c *awsDynamoDBScaler) GetMetrics(ctx context.Context, metricName string, metricSelector labels.Selector) ([]external_metrics.ExternalMetricValue, error) {
+	metricValue, err := c.GetQueryMetrics()
+
```
Comment (zroubalik) [inline pkg/scalers/aws_dynamodb_scaler.go:167]:
```suggestion
```

File: pkg/scalers/aws_dynamodb_scaler.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+			return nil, fmt.Errorf("error parsing metadata targetValue")
+		}
+
+		meta.targetValue = n
+	} else {
+		return nil, fmt.Errorf("no targetValue given")
+	}
+
+	auth, err := getAwsAuthorization(config.AuthParams, config.TriggerMetadata, config.ResolvedEnv)
+
```
Comment (zroubalik) [inline pkg/scalers/aws_dynamodb_scaler.go:125]:
```suggestion
```

File: pkg/scalers/aws_dynamodb_scaler.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	targetValue               int
+	awsAuthorization          awsAuthorizationMetadata
+	scalerIndex               int
+}
+
+var dynamoDBLog = logf.Log.WithName("aws_dynamodb_scaler")
+
+func NewAwsDynamoDBScaler(config *ScalerConfig) (Scaler, error) {
+	meta, err := parseAwsDynamoDBMetadata(config)
+
```
Comment (zroubalik) [inline pkg/scalers/aws_dynamodb_scaler.go:57]:
```suggestion
```

File: pkg/scalers/aws_dynamodb_scaler.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+		return 0, err
+	}
+
+	return *res.Count, nil
+}
+
+// json2Map convert Json to map[string]string
+func json2Map(js string) (m map[string]*string, err error) {
+	err = bson.UnmarshalExtJSON([]byte(js), true, &m)
+
```
Comment (zroubalik) [inline pkg/scalers/aws_dynamodb_scaler.go:234]:
```suggestion
```

**Review by zroubalik — COMMENTED** [review]
Looking good, just a few minor nits on formatting.

Also a big topic is e2e testing. Do you think you can implement e2e tests for this scaler, similar to this https://github.com/kedacore/keda/tree/main/tests? We should have KEDA related AWS account soon CC @tomkerkhove

File: pkg/scalers/aws_dynamodb_scaler.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+	"fmt"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbiface"
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
```
[the remark this answers is not in the document]
Comment (samueleresca) [reply pkg/scalers/aws_dynamodb_scaler.go]: Now it should be fine

**Comment by samueleresca** [comment]
> Looking good, just a few minor nits on formatting.
> 
> Also a big ask is e2e testing. Do you think you can implement e2e tests for this scaler, similar to this https://github.com/kedacore/keda/tree/main/tests? We should have KEDA related AWS account soon CC @tomkerkhove

I'm happy to implement the e2e tests for this. Should I do it in this PR?

**Comment by zroubalik** [comment]
> > Looking good, just a few minor nits on formatting.
> > Also a big ask is e2e testing. Do you think you can implement e2e tests for this scaler, similar to this https://github.com/kedacore/keda/tree/main/tests? We should have KEDA related AWS account soon CC @tomkerkhove
> 
> I'm happy to implement the e2e tests for this. Should I do it in this PR?

Yes please :)

File: pkg/scalers/aws_dynamodb_scaler.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+	}
+
+	return append([]external_metrics.ExternalMetricValue{}, metric), nil
+}
+
+func (c *awsDynamoDBScaler) GetMetricSpecForScaling(context.Context) []v2beta2.MetricSpec {
+	targetMetricValue := resource.NewQuantity(int64(c.metadata.targetValue), resource.DecimalSI)
+	externalMetric := &v2beta2.ExternalMetricSource{
+		Metric: v2beta2.MetricIdentifier{
+			Name: GenerateMetricNameWithIndex(c.metadata.scalerIndex, kedautil.NormalizeString(fmt.Sprintf("aws-dynamodb-%s", c.metadata.tableName))),
```
Comment (zroubalik) [inline pkg/scalers/aws_dynamodb_scaler.go:184]: Could you please move this to `parseAwsDynamoDBMetadata()` so it is called only on the first time?

You can take a look here, to see the example:
https://github.com/kedacore/keda/blob/b6bc7b754e2b5145c45895126e9dc57e86ab8816/pkg/scalers/datadog_scaler.go#L134

**Comment by samueleresca** [comment]
> > > Looking good, just a few minor nits on formatting.
> > > Also a big ask is e2e testing. Do you think you can implement e2e tests for this scaler, similar to this https://github.com/kedacore/keda/tree/main/tests? We should have KEDA related AWS account soon CC @tomkerkhove
> > 
> > 
> > I'm happy to implement the e2e tests for this. Should I do it in this PR?
> 
> Yes please :)

Regarding the e2e testing. I'm considering spanning up an instance using the following image for testing purposes: https://hub.docker.com/r/amazon/dynamodb-local. Until we don't have an AWS account. It should mimic the dynamodb on AWS, except for the fact that multiple regions are not supported

**Comment by samueleresca** [comment]
@zroubalik  I updated the PR with the tests.  I executed the E2E test locally with my personal subscription and looks good to me. Let me know when you have the account available.

**Comment by zroubalik** [comment]
> @zroubalik I updated the PR with the tests. I executed the E2E test locally with my personal subscription and looks good to me. Let me know if I can do something else.

@JorTurFer @tomkerkhove how should we approach this? Do we know if there's any progress on the AWS account from CNCF?

File: tests/scalers/aws-dynamodb.test.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+  dynamoClient = new DynamoDBClient({
+    region: awsRegion,
+    credentials: {
+      accessKeyId: awsAccessKey,
+      secretAccessKey: awsSecretKey
+    }
+  });
+
+  // create namespace
+  sh.exec(`kubectl create namespace ${dynamoDBNamespace}`)
```
Comment (zroubalik) [inline tests/scalers/aws-dynamodb.test.ts:39]: There's a helper method for creating namespace. please use it.

**Review by zroubalik — CHANGES_REQUESTED** [review]
Nit

**Comment by JorTurFer** [comment]
> > @zroubalik I updated the PR with the tests. I executed the E2E test locally with my personal subscription and looks good to me. Let me know if I can do something else.
> 
> @JorTurFer @tomkerkhove how should we approach this? Do we know if there's any progress on the AWS account from CNCF?

If the e2e test is running, I'd keep it commented and I'd enable it when the account is ready. We already have other scalers without any e2e test, so I wouldn't block this PR. 
Other option is using [`DynamoDB Local`](https://hub.docker.com/r/amazon/dynamodb-local) but I'm not sure if that development version could be enough for our e2e test case. Even if the local version is enough, I'd prefer using the real version, but this is an option

**Comment by zroubalik** [comment]
> > > @zroubalik I updated the PR with the tests. I executed the E2E test locally with my personal subscription and looks good to me. Let me know if I can do something else.
> > 
> > 
> > @JorTurFer @tomkerkhove how should we approach this? Do we know if there's any progress on the AWS account from CNCF?
> 
> If the e2e test is running, I'd keep it commented and I'd enable it when the account is ready. We already have other scalers without any e2e test, so I wouldn't block this PR. Other option is using [`DynamoDB Local`](https://hub.docker.com/r/amazon/dynamodb-local) but I'm not sure if that development version could be enough for our e2e test case. Even if the local version is enough, I'd prefer using the real version, but this is an option

Yeah, let's disable e2e tests for now and enable once we have the account.

**Comment by tomkerkhove** [comment]
No AWS account yet indeed. I will check

File: CHANGELOG.md [hunk 8]
```diff
@@ -30,6 +30,9 @@
 
 ### New
 
+- Add DynamoDB Scaler ([#2486](https://github.com/kedacore/keda/pull/2486))
+- **General:** Automatically release container image for ARM ([#2263]https://github.com/kedacore/keda/issues/2263))
+- **General:** Automatically run end-to-end tests on ARM ([#2262]https://github.com/kedacore/keda/issues/2262))
```
Comment (zroubalik) [inline CHANGELOG.md:35]: Could you please include just the relevant info? and use the format as it is used for other new scalers (please also include AWS in the name)

File: tests/scalers/aws-dynamodb.test.ts [hunk 9]
```diff
@@ -0,0 +1,204 @@
+import test from 'ava'
```
Comment (zroubalik) [inline tests/scalers/aws-dynamodb.test.ts:1]: Could you please comment the whole content of the file, so it is not being executed, we will enable it once we have AWS account in place.

**Review by JorTurFer — COMMENTED** [review]
You have done an awesome job! 🤟🤟🤟 
Thanks for this contribution and all the effort, LGTM
Only 1 small nit inline

**Review by JorTurFer — APPROVED** [review]
LGTM!
Thanks for the huge effort!

**Review by zroubalik — APPROVED** [review]
LGTM, thanks for the contribution! Great job.

File: CHANGELOG.md [hunk 10]
```diff
@@ -30,6 +30,9 @@
 
 ### New
 
+- **General:** Automatically release container image for ARM ([#2263]https://github.com/kedacore/keda/issues/2263))
```
Comment (zroubalik) [inline CHANGELOG.md:33]: I missed this, could you please remove the unrelated entries

**Review by zroubalik — CHANGES_REQUESTED** [review]

**Pull request merged**